Go 1.8 support (version bump to GopherJS 1.8-1).

shurcooL opened the pull request:
This is a WIP PR to track work to support Go 1.8. `go1.8beta1` has been released.

Each commit is bite-sized, with rationale and description, and fixes a specific issue.

It's now at a stage where only some of the standard library tests are failing. I'll make a table to track remaining work.

**Update:** All failing tests have been fixed, but there are some known TODOs remaining in code.

- [x] `database/sql/driver`
	<details>
		<summary>Test Failure Details</summary>
		<pre>--- FAIL: TestValueConverters (0.10s)
		types_test.go:77: test 20: driver.defaultConverter(driver.i(9)) = 9 (driver.i); want 9 (int64)
		types_test.go:77: test 23: driver.defaultConverter(driver.bs([1])) = [1] (driver.bs); want [1] ([]uint8)
	FAIL
	FAIL	database/sql/driver	0.623s</pre>
	</details>

- [x] `hash/crc32`

- [x] `math`

- [x] `io`

- [x] `reflect`

- [x] `runtime`

- [x] `net/rpc/jsonrpc`

	<details>
		<summary>Test Failure Details</summary>
		<pre>/home/ubuntu/gopherjs/test.342850936:4
	Error.stackTraceLimit=Infinity;var $global,$module;if(typeof window!=="undefined"){$global=window;}else if(typeof self!=="undefined"){$global=self;}else if(typeof global!=="undefined"){$global=global;$global.require=require;}else{$global=this;}if($global===undefined||$global.Array===undefined){throw new Error("no global object found");}if(typeof module!=="undefined"){$module=module;}var $packages={},$idCounter=0;var $keys=function(m){return m?Object.keys(m):[];};var $flushConsole=function(){};var $throwRuntimeError;var $throwNilPointerError=function(){$throwRuntimeError("invalid memory address or nil pointer dereference");};var $call=function(fn,rcvr,args){return fn.apply(rcvr,args);};var $makeFunc=function(fn){return function(){return $externalize(fn(this,new($sliceType($jsObjectPtr))($global.Array.prototype.slice.call(arguments,[]))),$emptyInterface);};};var $mapArray=function(array,f){var newArray=new array.constructor(array.length);for(var i=0;i<array.length;i++){n

	Error: runtime error: native function not implemented: sync.runtime_SemacquireMutex
	    at $callDeferred (/home/ubuntu/gopherjs/test.342850936:4:28973)
	    at $panic (/home/ubuntu/gopherjs/test.342850936:4:29597)
	    at $throwRuntimeError (/home/ubuntu/gopherjs/test.342850936:7:3381)
	    at $packages.sync.AE (/home/ubuntu/gopherjs/test.342850936:11:12427)
	    at $packages.sync.Q.ptr.Lock (/home/ubuntu/gopherjs/test.342850936:11:9096)
	    at $packages.net/rpc.AL.ptr.sendResponse (/home/ubuntu/gopherjs/test.342850936:93:29773)
	    at $packages.net/rpc.AI.ptr.call (/home/ubuntu/gopherjs/test.342850936:93:32372)
	    at f (/home/ubuntu/gopherjs/test.342850936:4:1231)
	    at $goroutine (/home/ubuntu/gopherjs/test.342850936:4:30195)
	    at Timeout.$runScheduled [as _onTimeout] (/home/ubuntu/gopherjs/test.342850936:4:30964)
	    at tryOnTimeout (timers.js:224:11)
	    at Timer.listOnTimeout (timers.js:198:5)
	FAIL	net/rpc/jsonrpc	1.668s</pre>
	</details>

- [x] `sync`

	<details>
		<summary>Test Failure Details</summary>
		<pre>--- FAIL: TestMutexMisuse (0.01s)
	/home/ubuntu/gopherjs/test.238107529:4
	Error.stackTraceLimit=Infinity;var $global,$module;if(typeof window!=="undefined"){$global=window;}else if(typeof self!=="undefined"){$global=self;}else if(typeof global!=="undefined"){$global=global;$global.require=require;}else{$global=this;}if($global===undefined||$global.Array===undefined){throw new Error("no global object found");}if(typeof module!=="undefined"){$module=module;}var $packages={},$idCounter=0;var $keys=function(m){return m?Object.keys(m):[];};var $flushConsole=function(){};var $throwRuntimeError;var $throwNilPointerError=function(){$throwRuntimeError("invalid memory address or nil pointer dereference");};var $call=function(fn,rcvr,args){return fn.apply(rcvr,args);};var $makeFunc=function(fn){return function(){return $externalize(fn(this,new($sliceType($jsObjectPtr))($global.Array.prototype.slice.call(arguments,[]))),$emptyInterface);};};var $mapArray=function(array,f){var newArray=new array.constructor(array.length);for(var i=0;i<array.length;i++){n

	Error: runtime error: native function not implemented: syscall.runtime_BeforeFork
	    at $callDeferred (/home/ubuntu/gopherjs/test.238107529:4:28973)
	    at $panic (/home/ubuntu/gopherjs/test.238107529:4:29597)
	    at $b (/home/ubuntu/gopherjs/test.238107529:32:67954)
	    at $callDeferred (/home/ubuntu/gopherjs/test.238107529:4:29144)
	    at $panic (/home/ubuntu/gopherjs/test.238107529:4:29597)
	    at $throwRuntimeError (/home/ubuntu/gopherjs/test.238107529:7:3352)
	    at $packages.syscall.AN (/home/ubuntu/gopherjs/test.238107529:13:19394)
	    at $packages.syscall.AP (/home/ubuntu/gopherjs/test.238107529:13:20425)
	    at $packages.syscall.BC (/home/ubuntu/gopherjs/test.238107529:13:30325)
	    at Object.$packages.syscall.BE [as StartProcess] (/home/ubuntu/gopherjs/test.238107529:13:32168)
	    at $packages.os.AO (/home/ubuntu/gopherjs/test.238107529:16:14753)
	    at Object.$packages.os.AN [as StartProcess] (/home/ubuntu/gopherjs/test.238107529:16:11036)
	    at $packages.os/exec.M.ptr.Start (/home/ubuntu/gopherjs/test.238107529:29:17373)
	    at $packages.os/exec.M.ptr.Run (/home/ubuntu/gopherjs/test.238107529:29:12460)
	    at $packages.os/exec.M.ptr.CombinedOutput (/home/ubuntu/gopherjs/test.238107529:29:23335)
	    at $packages.sync_test.AU (/home/ubuntu/gopherjs/test.238107529:34:34594)
	    at $packages.testing.CG (/home/ubuntu/gopherjs/test.238107529:32:69409)
	    at $goroutine (/home/ubuntu/gopherjs/test.238107529:4:30195)
	    at Timeout.$runScheduled [as _onTimeout] (/home/ubuntu/gopherjs/test.238107529:4:30964)
	    at tryOnTimeout (timers.js:224:11)
	    at Timer.listOnTimeout (timers.js:198:5)
	FAIL	sync	0.705s</pre>
	</details>

shurcooL commented:
Looking at fixing `reflect` package next. It's pretty critical, and fixing it first might resolve some other problems (e.g., the `database/sql` failing test is likely a consequence of `reflect.DeepEqual` having a problem).

It currently fails without much output in CI:

```
$ gopherjs test --short --minify --run='^Test' reflect
FAIL	reflect	0.923s
```

Adding `-v` makes me see the problem is in `TestDeepEqual`:

```
$ gopherjs test -v --run=TestDeepEqual reflect
=== RUN   TestDeepEqual
FAIL	reflect	0.718s
```

Also, we may need to add support for the new [`reflect.Swapper`](https://tip.golang.org/pkg/reflect/#Swapper). It seems the standard library implementation only has an implementation that requires use of `unsafe` package (see https://tip.golang.org/src/reflect/swapper.go). But we can use the "safe" one from https://github.com/camlistore/go4/blob/master/reflectutil/swapper.go (huge thanks to Brad for implementing that version).

shurcooL commented:
Narrowed down the cause of `TestDeepEqual` to not completing. It's these new test cases, added in 1.8:

```

var deepEqualTests = []DeepEqualTest{
	// ... all the old ones continue to work okay.
	// but the new "Possible loops" cause trouble.
	// Commenting them out causes TestDeepEqual to pass.

	// Possible loops.
	{&loop1, &loop1, true},
	{&loop1, &loop2, true},
	{&loopy1, &loopy1, true},
	{&loopy1, &loopy2, true},
}
```

After that, only `TestSwapper` needs to be made to pass, and all of `reflect` tests succeed.

shurcooL commented:
More specifically, it's only these cases that are problematic:

```
// Possible loops.
{&loop1, &loop1, true}, // ok
{&loop1, &loop2, true}, // fails
{&loopy1, &loopy1, true}, // ok
{&loopy1, &loopy2, true}, // fails
```

And `TestSwapper` passes with a safe swapper implementation.

shurcooL commented:
Great, CI is passing now! ✅ 

Next up is addressing remaining TODOs.

Most of the failures that I fixed weren't regressions in behavior with 1.8, but rather known/existing issues being triggered. I didn't have to do much to get previously working things to continue to work.

But first, I will rewrite history and squash some related commits to make the PR smaller and each commit more logical.

shurcooL commented:
Ok, I've done an interactive rebase to squash/re-arrange commits and clean up the history to have more logical, well separated changes in each commit.

Since this is now passing CI, do you want to look over it @neelance and leave any first-order comments that you might have?

Next up we should look at the remaining TODOs inside the code and see if we can address them.

That said, for anyone eager to try this `go1.8` branch and test it out, feel welcome to do that. Report any issues either here (for small things) or as issues (label with go1.8 as needed).

neelance commented on build/build.go:
```diff
@@ -90,8 +90,6 @@ func importWithSrcDir(path string, srcDir string, mode build.ImportMode, install
 	case "crypto/x509":
 		pkg.CgoFiles = nil
-	case "hash/crc32":
-		pkg.GoFiles = []string{"crc32.go", "crc32_generic.go"}
```
Yey.

neelance commented on compiler/natives/src/math/math_test.go:
```diff
@@ -0,0 +1,12 @@
+//       should override TestGamma specifically and not the package-wide tolerances,
+//       because this will cause many other tests to be less accurate. Or maybe this
+//       is fine?
+func close(a, b float64) bool     { return tolerance(a, b, 4e-14) }
```
I'm still wondering where exactly we're losing precision. I guess it is because we're using some JS native math functions which are implemented slightly differently than the Go ones. Investigate?

neelance replied:
Any idea what's going on here?

shurcooL replied:
Made #589 to track this.

neelance commented on compiler/natives/src/reflect/swapper.go:
```diff
@@ -0,0 +1,29 @@
+
+package reflect
+
+func Swapper(slice interface{}) func(i, j int) {
```
`reflect.Swapper`, interesting addition. I guess this is because of `sort`?

shurcooL replied:
Yeah, this is to enable the new way of sorting in 1.8.

https://beta.golang.org/pkg/sort/#example_Slice

neelance commented on tool.go:
```diff
@@ -921,22 +921,8 @@ var examples = []testing.InternalExample{
-
 func main() {
-	m := testing.MainStart(matchString, tests, benchmarks, examples)
+	m := testing.MainStart(testdeps.TestDeps{}, tests, benchmarks, examples)
```
How is `matchString` implemented now? As far as I remember this was done to avoid the dependency from `testing` to `regexp`.

shurcooL replied:
The commit message of 3b29828ba7778097142fd8db96ecf75d9d3a341c says:

> See https://github.com/golang/go/commit/c56cc9b3b5727647c2afb3d57f5793151558a0a7?w=1.

That commit shows that it's still implemented the same way, using `regexp`. They've just shuffled imports around into `testdeps` package.

shurcooL commented:
One minor issue I believe I'm seeing on 1.8 branch is that there's the following warning showing up:

![image](https://cloud.githubusercontent.com/assets/1924134/22436111/b4b5fada-e6f0-11e6-9b32-7e65879926ca.png)

In situations where it shouldn't be. I don't think it was showing up in 1.7. This is not fully confirmed yet.

shurcooL commented:
I've rebased this on top of latest `master` and cleaned up the history (got rid of multiple Go beta version update commits, combined all the "regenerate" commits into one, etc.).

I see there's one failure:

```
unok	fixedbugs/issue8613.go	0.103s
```

That means a previously known failure is now passing. It might be because of something that was fixed in `master` since then.

I'll need to look into that after I get back from the Go 1.8 release party! 🎉

shurcooL commented:
The unexpected passing test above makes perfect sense. Here it is:

```
"fixedbugs/issue8613.go":  {desc: "Does not panic upon division by zero when assigned to _, see https://github.com/gopherjs/gopherjs/issues/551."},
```

And https://github.com/gopherjs/gopherjs/issues/551 is now resolved (on master).

So that needs to be updated to a "known pass".

lologarithm commented:
+1 1.8 support looking good on my projects!

shurcooL commented:
GopherJS 1.8-1 is now released and available on `master` branch.

See blog post about it here: https://twitter.com/GopherJS/status/832708146443399170.

See tweet here: https://twitter.com/GopherJS/status/832708146443399170.

Thanks everyone! 🎉
